lib/Graph/Graph.cpp

```diff
@@ -488,6 +502,32 @@ static void assertConvDims(NodeValue input, NodeValue filter, NodeValue bias,
   assert(bias.getType()->size() == filterDims[0] && "Invalid bias size");
 }
 
+static void assertConv3DDims(NodeValue input, NodeValue filter, NodeValue bias,
+                             llvm::ArrayRef<unsigned_t> kernels,
+                             llvm::ArrayRef<unsigned_t> strides,
+                             llvm::ArrayRef<unsigned_t> pads,
+                             unsigned_t group) {
+
+  ShapeNHWCT idim(input.dims());
+  ShapeHWT kdim(kernels);
+  (void)kdim;
+  check3DKernelSize(idim, kernels, pads);
+  assert(idim.c % group == 0 && "channels number must be divisible by groups");
+
+  // NOTE: here the N in NHWCT is abnormal because it is the number of filters
```

Comment:
I'm a little confused -- what is functionally different here for the logic in here? It seems to be the same as `assertConvDims()` except with the time dim, no?

Reply:
The functionality here is the same except with the extra dimension. The note is referring to the fact that normally in things with shape NHWCT the N dimension is the number of examples but because this is being used on the filter, N is actually the number of output channels of the convolution which is different. The rest of the channels HWCT are all the same as what you'd expect them to be.

Reply:
Ah ok, so then this note is applicable to 2D convs too -- this was a little confusing of a note since it's only in `assertConv3DDims()`, because I thought it was calling out a difference between 2D and 3D. Maybe makes sense to call out specifically that there's no difference, or include the note in the 2D case, or move the note to somewhere else?

Reply:
@jfix71 it doesn't apply to the 2D implementation because, although the fliters are laid out the same, the 2D implementation doesn't choose to use of `NHWC` and instead indexes directly into the shape vectors. I think that way is less readable so I used `NHWCT` for the 3D case even though it's technically not the correct usage of `NHWCT` beacuse the N dimension is different than normal which is what I noted here. I think I'll just changes the 2D implementation to use `NHWC` and add the same note there as well since it's just easier to read and understand what's going on because you can understand what each dimension of the filters tensor is.

Reply:
Ahh ok got it, understood, thanks!